### DISCUSSION TITLE
Workflow/Lifecycle stages feature like Pipefy, Jira, Hubspot, ServiceNow, Salesforce

### DISCUSSION THREAD

**Discussion by manjeshpv** [twentyhq/twenty#6991] [Ideas]
- Jira and Hubspot have strong workflows

**Comment by greptile-apps** [comment]
To implement a workflow/lifecycle stages feature similar to Pipefy, Jira, and Hubspot, follow these steps:

1. **Define Workflow Stages**: Extend the existing Kanban functionality to support custom workflow stages. Modify the `Select Field Settings` to allow users to define stages.

2. **Update Kanban Views**: Enhance the Kanban views to support the new workflow stages. Refer to `/packages/twenty-website/src/content/user-guide/objects/kanban-views.mdx` for existing Kanban view functionalities.

3. **Database Schema**: Update the database schema to store workflow stages. Modify the migration files in `/packages/twenty-website/src/database/migrations/meta` to include a new table for workflow stages.

4. **API Integration**: Update the API routes to handle CRUD operations for workflow stages. Modify `/packages/twenty-website/src/app/contributors/api/generate/route.tsx` to include endpoints for managing workflow stages.

5. **UI Components**: Create new UI components for managing workflow stages. Add these components to the relevant pages and integrate them with the existing Kanban board.

6. **Testing**: Ensure that the new workflow stages feature is thoroughly tested. Update the GitHub Actions workflows in `/.github/workflows` to include tests for the new feature.


### References
[/.github/workflows](https://github.com/twentyhq/twenty/blob/main/.github/workflows)
[/packages/twenty-website/src/github-sync/contributors/types.tsx](https://github.com/twentyhq/twenty/blob/main/packages/twenty-website/src/github-sync/contributors/types.tsx)
[/packages/twenty-website/src/app/contributors/api/generate/route.tsx](https://github.com/twentyhq/twenty/blob/main/packages/twenty-website/src/app/contributors/api/generate/route.tsx)
[/.github/ISSUE_TEMPLATE/add-new-feature---behavior.md](https://github.com/twentyhq/twenty/blob/main/.github/ISSUE_TEMPLATE/add-new-feature---behavior.md)
[/.github/ISSUE_TEMPLATE/request-technical-work.md](https://github.com/twentyhq/twenty/blob/main/.github/ISSUE_TEMPLATE/request-technical-work.md)
[/.github/ISSUE_TEMPLATE/report-bug.md](https://github.com/twentyhq/twenty/blob/main/.github/ISSUE_TEMPLATE/report-bug.md)
[/packages/twenty-website/src/content/user-guide/other/glossary.mdx](https://github.com/twentyhq/twenty/blob/main/packages/twenty-website/src/content/user-guide/other/glossary.mdx)
[/packages/twenty-website/src/content/releases/0.3.3.mdx](https://github.com/twentyhq/twenty/blob/main/packages/twenty-website/src/content/releases/0.3.3.mdx)
[/packages/twenty-website/src/content/user-guide/objects/kanban-views.mdx](https://github.com/twentyhq/twenty/blob/main/packages/twenty-website/src/content/user-guide/objects/kanban-views.mdx)
[/packages/twenty-website/src/database/migrations/meta](https://github.com/twentyhq/twenty/blob/main/packages/twenty-website/src/database/migrations/meta)

<details>
<summary>

#### About Greptile
</summary>

This response provides a starting point for your research, not a precise solution.

Help us improve! Please leave a 👍 if this is helpful and 👎 if it is irrelevant.

[Ask Greptile](https://app.greptile.com/chat/github/twentyhq/twenty/main) · [Edit Issue Bot Settings](https://app.greptile.com/apps/github)
</details>

**Comment by Bonapara** [comment]
Hi @manjeshpv, we began working on workflows this week!

**Comment by manjeshpv** [comment]
can you share details, i can also contribute

i thought of developing npm package, which can be used across twenty or any other projects

Features i wanted
- Users CRUD - _already exists in twenty_
- Stages CRUD - _I created custom objects_
- Pipelines CRUD (Pipelines is categorizing stages) - _I created custom objects_
- Teams or Roles CRUD
- Building state machine with state charts (https://www.youtube.com/watch?v=OSFYcfbvdpg&t=25s)

Nice to have:
- Resource Groups - Territory Management or Segmentation wise Access Controls
- Reports
- Recommendations emails about outdated stages
- Triggers/Rules/TTL to automatically close the tickets

Lets take a real life example:

In our company we configure a storage server for every school, 

**Pipeline 1: Warehouse**
- Multiple stages involved in warehouse pipeline like 
    - Assembling Hardware
    - Configuration
    - Data sync
    - Quality checks
then shipped via logistic partner

**Pipeline 2: Servers in School & Healthy**
- School receives the the server  and informs IT Support
- Our central IT support guy will schedule a visit by IT service guy in the location 
- Local guy visits school and 
  - unbox and configure the internet router etc
  - do quality checks to verify the server is connecting with school teachers android and ios tablets
  - meanwhile remotely our IT team ensures the server sending monitoring events & Health data to central control panel

**Pipeline 3: Support Flow - Unhealthy**
- If any problem happens to any staff in school while connecting with storage server, they will call our company client relationship manager or IT Support
- IT Team move school record in twenty crm to support pipeline
- In support pipeline, their are multiple stages like 
   - Router Issue
   - Internet issue
   - Disk failure
 
Like this we given access to all our employees and school IT guy to Twenty CRM, we have around 5k users monthly active users. everyone interact with CRM to collaborate and finish their day today work

We have hierarchy of territory managers like (Zonal, Regional, 28 State, Per school) all this people like to see the CRM by default loads schools associated with their Designation/Role/Assigned schools/Devices


https://github.com/nocodb/nocodb/issues/8980

**Comment by manjeshpv** [comment]
![image](https://github.com/twentyhq/twenty/assets/1999957/937ded84-cbc1-43d5-88bf-6310d6e36e49)
https://github.com/makeplane/plane/issues/5250

**Comment by greptile-apps** [comment]
To implement a workflow/lifecycle stages feature similar to Pipefy, Jira, and Hubspot, follow these steps:

1. **Kanban Views**: Extend the existing Kanban functionality.
   - File: `/packages/twenty-website/src/content/user-guide/objects/kanban-views.mdx`
   - Add functionality to customize stages and transitions.

2. **Workflow Automation**: Integrate with GitHub Actions for automated workflows.
   - Directory: `/.github/workflows`
   - Define new workflows for lifecycle stages.

3. **Release Notes**: Document new features in release notes.
   - File: `/packages/twenty-website/src/content/releases/0.3.3.mdx`
   - Add a section for the new workflow/lifecycle stages feature.

Example code snippets and detailed steps can be added to the respective files to ensure seamless integration.


### References
[/.github/workflows](https://github.com/twentyhq/twenty/blob/main/.github/workflows)
[/packages/twenty-website/src/content/user-guide/objects/kanban-views.mdx](https://github.com/twentyhq/twenty/blob/main/packages/twenty-website/src/content/user-guide/objects/kanban-views.mdx)
[/packages/twenty-website/src/content/releases/0.3.3.mdx](https://github.com/twentyhq/twenty/blob/main/packages/twenty-website/src/content/releases/0.3.3.mdx)
[/packages/twenty-website/src/content/releases/0.2.3.mdx](https://github.com/twentyhq/twenty/blob/main/packages/twenty-website/src/content/releases/0.2.3.mdx)

<details>
<summary>

#### About Greptile
</summary>

This response provides a starting point for your research, not a precise solution.

Help us improve! Please leave a 👍 if this is helpful and 👎 if it is irrelevant.

[Ask Greptile](https://app.greptile.com/chat/github/twentyhq/twenty/main) · [Edit Issue Bot Settings](https://app.greptile.com/apps/github)
</details>

**Comment by BOHEUS** [comment]
Duplicate of https://github.com/twentyhq/twenty/discussions/10631
